### ISSUE TITLE
Feature request: recursively check projects down the dependency tree

### ISSUE THREAD

**Issue by sunshineo** [draperunner/ditched#1]
Thank you for saving the abandoned abandoned project!
I think if we add a "-r" option to go down the dependency tree recursively, it can reveal a lot more useful info

**Comment by draperunner** [comment]
Hi! Thanks for the feature request!

So if I understand you correctly, you want to list all the ditched _transitive_ dependencies? In other words, all the packages that are ditched that you don't have in your own package.json, but that some other package depends on (and then recursively down the tree). 

First of all, I'm not sure how useful this is. You cannot really do something if your (newly released) dependency `a` is dependent on a ditched package `old`. Or what use cases were you thinking of?

Also, some design updates would be needed if this should be implemented. I guess you would want to know which dependencies are dependent on the package, like `a -> b -> c -> ditched_pkg`. Maybe a new column is sufficient, or some sort of tree view.

**Comment by sunshineo** [comment]
You are right that if the dependency is not owned by me, I cannot fix it. But I do make a lot of small packages and use those in a bigger project. I can run ditched in each package project one by one, not a big problem.
Also for packages I do not own, I can talk to the owner on Github to try to let them switch away from ditched dependencies
Right now I have to clone/fork the dependency and run ditched in it

For the UI, maybe we can look at `npm list --depth=2`. I don't think we have to stick to 3 vertical aligned columns with headers, a tab between should be OK. The data for the 2nd column is clearly latest release. And the third column instead of yes/no, we can write 'ditched'/'active
![image](https://user-images.githubusercontent.com/1072766/171701579-d279cb4b-4f26-4a64-89c2-57a3a7d311f9.png)

**Comment by draperunner** [comment]
I like it! I suggest we could start with adapting this `npm list`-like view with today's functionality. Then add the recursive feature.

If you want to have a go at this, I'm accepting PRs! 😄

**Comment by sunshineo** [comment]
I'll try.
But going for this new UI without the recursive will be ugly. I would like to actually try to go for recursive using the old UI. It will be confusing without the indentation but if you do not use the `--depth=x` flag then it won't do the recursive thing

**Comment by sunshineo** [comment]
I got the recursive working but 2 issues:
When goes down multiple levels, it will be a linear process and slow
We cannot do a tree view yet, especially if -a is not present, you cannot get a proper tree

**Comment by draperunner** [comment]
Great work on the PR! Some thoughts:

I'm not sure that the new UI would look so bad with today's shallow functionality, though. `npm list` looks OK, doesn't it? I guess if we wanted to align the tab-separated columns, that could require some more work. The `npm list`-like view is also more script-friendly, if someone would want to pipe the results to another script. I think at least the two views should be similar, having two entirely different UIs depending on the `level` parameter seems a bit weird.

If we are to use the old table view, I think we should display some info about the dependency relations. Perhaps a new column "Depended on by" which shows the parent, or some indentation within the first column.

It makes sense that the recursive functionality will be much slower. But what if we use the local `node_modules` to build the dependency tree beforehand, then fetch every dependency in chunks, parallelized? We could read the output of `npm list` (easiest) or read the contents of node_modules ourselves. This would be much faster, but would of course require your to have run `npm install` in your project before running `ditched`. I think that's fair.

**Comment by sunshineo** [comment]
If we remove the existing `-a` option, then the `npm list` tree UI makes much more sense compare to the current table UI for supporting recursive. I do not have a good idea on how to support `-a` and recursive with the tree UI.

It is possible to use the data installed under `node_modules`, but the difference is the installed version could be old and we falsely report the project is abandoned.

**Comment by draperunner** [comment]
Completed in #3
